graph/run-dynamic.go

```diff
@@ -263,59 +264,71 @@ func ConvergeDynamicThread[EP EPP[E], V VPI[V], E EPI[E], M MVI[M], N any, A Alg
 				// There is a request for synchronously syncing topology (blocking), acknowledge here as we have no more topology events to process.
 				bspSync = false
 				gt.Response <- ACK
-			} else if remitCount == 0 {
-				// Want topology events, but none found -- we should relax.
-				topFail++
-				gt.Status = BACKOFF_TOP
-				utils.BackOff(topFail)
-				if timeStates && tidx == 0 {
-					curr = time.Now()
-					gt.LoopTimes[5] += curr.Sub(prev)
-					prev = curr
-				}
-			} else {
-				topFail = 0
+			} else if remitCount == 0 && !epoch {
+				topFailed = true
 			}
 
 			if strucClosed {
 				doneEvents <- struct{}{}
-			} else if TOPOLOGY_FIRST || blockAlgIfTop {
-				continue
 			}
 		}
 
-		// Process algorithm messages. Check for algorithm termination if needed.
-		gt.Status = APPLY_MSG
-		checkTerm := (strucClosed && remitClosed) || (epoch && remitCount == 0 && count == 0)
-		completed, count = ProcessMessages[V, E, M, N](alg, g, gt, checkTerm)
-		if completed && epoch {
-			//log.Debug().Msg("T[" + utils.F("%02d", tidx) + "] completed epoch")
-			gt.Status = DONE
-			gt.Response <- ACK
-
-			resp := <-gt.Command // BLOCK and wait for resume
-			if resp != RESUME {
-				log.Panic().Msg("Expected to resume after blocked")
+		processAlg := strucClosed || !(TOPOLOGY_FIRST || blockAlgIfTop)
+		algFailed := false
+		if processAlg {
+			// Process algorithm messages. Check for algorithm termination if needed.
+			gt.Status = APPLY_MSG
+			checkTerm := (strucClosed && remitClosed) || (epoch && remitCount == 0 && topCount == 0)
+			completed, algCount = ProcessMessages[V, E, M, N](alg, g, gt, checkTerm)
+			if completed && epoch {
+				//log.Debug().Msg("T[" + utils.F("%02d", tidx) + "] completed epoch")
+				gt.Status = DONE
+				gt.Response <- ACK
+
+				resp := <-gt.Command // BLOCK and wait for resume
+				if resp != RESUME {
+					log.Panic().Msg("Expected to resume after blocked")
+				}
+				epoch = false
+				completed = false
+			} else if algCount == 0 {
+				algFailed = true
 			}
-			epoch = false
-			completed = false
-			g.TerminateVotes[tidx] = 0
 		}
 
-		if count == 0 {
-			algFail++
-			if algFail%10 == 0 {
-				gt.Status = BACKOFF_ALG
-				utils.BackOff(algFail / 10)
+		// Backoff only when there are no topological events and algorithmic event
+		//if (topFailed && (!processAlg || algFailed)) || (algFailed && (!processTop || topFailed)) {
+		if (topFailed && (!processAlg || algFailed)) || (algFailed && (topFailed)) {
```

Comment:
Replacing the condition on Line 301 with the condition on Line 300 slows the PageRank test down by 50%

Reply:
We can also consider using the condition `(!processTop || topFailed) && (!processAlg || algFailed)`, which has a different behaviour when both `processTop` and `processAlg` are false.

Reply:
Note that the conditional: ```(topFailed && (!processAlg || algFailed)) || (algFailed && (topFailed))``` 
simplifies to just ```(topFailed && (!processAlg || algFailed))```

Reply:
I think ```(!processTop || topFailed) && (!processAlg || algFailed)``` makes the most sense for now.
I wouldn't worry about performance on the tiny go test cases.

Ideally we should make a "standard" set of tasks (differing algorithms and async/dynamic and query counts) with a medium graph, and make sure that framework-level changes don't degrade performance for them.

Reply:
I have change it too `(!processTop || topFailed) && (!processAlg || algFailed)`. 

Yeah it would definitely be nice to have a set of benchmarks :)